server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoGridTiler.java

```diff
@@ -0,0 +1,256 @@
+     * @param y        the y-coordinate
+     * @param precision  the zoom level of tiles
+     */
+    long encode(double x, double y, int precision);
+
+    /**
+     *
+     * @param docValues  the array of long-encoded bucket keys to fill
+     * @param geoValue   the input shape
+     * @param precision  the tile zoom-level
+     *
+     * @return the number of tiles the geoValue intersects
+     */
+    int setValues(CellIdSource.GeoShapeCellValues docValues, MultiGeoValues.GeoValue geoValue, int precision);
+
+    class GeoHashGridTiler implements GeoGridTiler {
+        public static final GeoHashGridTiler INSTANCE = new GeoHashGridTiler();
+
+        private GeoHashGridTiler() {}
+
+        @Override
+        public long encode(double x, double y, int precision) {
+            return Geohash.longEncode(x, y, precision);
+        }
+
+        @Override
+        public int setValues(CellIdSource.GeoShapeCellValues values, MultiGeoValues.GeoValue geoValue, int precision) {
+            MultiGeoValues.BoundingBox bounds = geoValue.boundingBox();
+            assert bounds.minX() <= bounds.maxX();
+            long numLonCells = (long) ((bounds.maxX() - bounds.minX()) / Geohash.lonWidthInDegrees(precision));
+            long numLatCells = (long) ((bounds.maxY() - bounds.minY()) / Geohash.latHeightInDegrees(precision));
+            long count = (numLonCells + 1) * (numLatCells + 1);
+            if (count == 1) {
+                String hash = Geohash.stringEncode(bounds.minX(), bounds.minY(), precision);
+                values.resizeCell(1);
+                values.add(0, Geohash.longEncode(hash));
+                return 1;
+            } else if (count <= precision) {
+                return setValuesByBruteForceScan(values, geoValue, precision, bounds);
+            } else {
+                return setValuesByRasterization("", values, 0, precision, geoValue, bounds);
+            }
+        }
+
+        protected int setValuesByBruteForceScan(CellIdSource.GeoShapeCellValues values, MultiGeoValues.GeoValue geoValue,
+                                                int precision, MultiGeoValues.BoundingBox bounds) {
+            // TODO: This way to discover cells inside of a bounding box seems not to work as expected. I  can
+            // see that eventually we will be visiting twice the same cell which should not happen.
+            int idx = 0;
+            String min = Geohash.stringEncode(bounds.minX(), bounds.minY(), precision);
+            String max = Geohash.stringEncode(bounds.maxX(), bounds.maxY(), precision);
+            double minY = Geohash.decodeLatitude(min);
+            double minX = Geohash.decodeLongitude(min);
+            double maxY = Geohash.decodeLatitude(max);
+            double maxX = Geohash.decodeLongitude(max);
+            for (double i = minX; i <= maxX; i += Geohash.lonWidthInDegrees(precision)) {
+                for (double j = minY; j <= maxY; j += Geohash.latHeightInDegrees(precision)) {
```

Comment:
I wonder whether accuracy issues of floating-point additions could cause bugs here, is it possible to work directly on the encoded space instead?

Reply:
this is a good point. I can investigate this after the merge maybe? I have added quiet a bit of tests that do not show this to be a problem